In [1]:
import sys

In [2]:
from time import time

In [3]:
sys.path.append("../tools/")

In [4]:
from email_preprocess import preprocess

In [5]:
### features_train and features_test are the features for the training
### and testing datasets, respectively
### labels_train and labels_test are the corresponding item labels

In [19]:
feature_train, feature_test, label_train, label_test = preprocess()

no. of Chris training emails: 7936
no. of Sara training emails: 7884


In [7]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

In [9]:
t0 = time()
clf = SVC(kernel = "linear")
print "starting time:", round(time()-t0, 3), "s"
clf.fit(feature_train, label_train)
print "training time:", round(time()-t0, 3), "s"
prediction = clf.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print accuracy
print "predicting time:", round(time()-t0, 3), "s"

starting time: 0.007 s
training time: 197.899 s
0.984072810011
predicting time: 218.411 s


### What is the accuracy of the classifier?

0.98407281001

### What is the time for classify

    197.899 s     218.411 s

##  Train with less data

In [20]:
feature_train1 = feature_train[:len(feature_train)/100] 
label_train1 = label_train[:len(label_train)/100] 
t0 = time()
clf = SVC(kernel = "linear")
print "starting time:", round(time()-t0, 3), "s"
clf.fit(feature_train1, label_train1)
print "training time:", round(time()-t0, 3), "s"
prediction = clf.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print accuracy
print "predicting time:", round(time()-t0, 3), "s"

starting time: 0.001 s
training time: 0.124 s
0.884527872582
predicting time: 1.419 s


## train with less data but more complex model

In [21]:
feature_train2 = feature_train[:len(feature_train)/100] 
label_train2 = label_train[:len(label_train)/100] 
t0 = time()
## model complex due to the new kernel rbf - radial basis functions
clf = SVC(kernel = "rbf")
print "starting time:", round(time()-t0, 3), "s"
clf.fit(feature_train2, label_train2)
print "training time:", round(time()-t0, 3), "s"
prediction = clf.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print accuracy
print "predicting time:", round(time()-t0, 3), "s"

starting time: 0.0 s
training time: 0.138 s
0.616040955631
predicting time: 1.592 s


## train with less data but more complex model but different C values (10.0, 100., 1000., and 10000)

In [29]:
feature_train3 = feature_train[:len(feature_train)/100] 
label_train3 = label_train[:len(label_train)/100] 

## model complex due to the new kernel rbf - radial basis functions
clf1 = SVC(kernel = "rbf",C = 10)
clf1.fit(feature_train3, label_train3)
prediction = clf1.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print "when C = 10"
print accuracy

clf2 = SVC(kernel = "rbf",C = 100)
clf2.fit(feature_train3, label_train3)
prediction = clf2.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print "when C = 100"
print accuracy

clf3 = SVC(kernel = "rbf",C = 1000)
clf3.fit(feature_train3, label_train3)
prediction = clf3.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print "when C = 1000"
print accuracy

clf4 = SVC(kernel = "rbf",C = 10000)
clf4.fit(feature_train3, label_train3)
prediction = clf4.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print "when C = 10000"
print accuracy


when C = 10
0.616040955631
when C = 100
0.616040955631
when C = 1000
0.821387940842
when C = 10000
0.892491467577


## Optimize model accuracy

In [30]:
clf = SVC(kernel = "rbf",C = 10000)
clf.fit(feature_train, label_train)
prediction = clf.predict(feature_test)
accuracy = accuracy_score(prediction, label_test)
print "when C = 10000, use the whole data"
print accuracy

when C = 10000, use the whole data
0.990898748578


## Use 1% of data, C equal to 10000, 
## print the result of 10, 26, 50th in test
## (0/1corresponding to Sara / Chris respectively)

In [31]:
feature_train4 = feature_train[:len(feature_train)/100] 
label_train4 = label_train[:len(label_train)/100]
clf4 = SVC(kernel = "rbf",C = 10000)
clf4.fit(feature_train3, label_train3)
pred_10 = clf4.predict(feature_test[10])
pred_26 = clf4.predict(feature_test[26])
pred_50 = clf4.predict(feature_test[50])
print pred_10, pred_26, pred_50

[1] [0] [1]


C:\Users\xyang\Anaconda2\lib\site-packages\sklearn\utils\validation.py:386: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and willraise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)
C:\Users\xyang\Anaconda2\lib\site-packages\sklearn\utils\validation.py:386: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and willraise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)
C:\Users\xyang\Anaconda2\lib\site-packages\sklearn\utils\validation.py:386: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and willraise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning

## There are over 1700 test events--how many are predicted to be in the “Chris” (1) class? (Use the RBF kernel, C=10000., and the full training set.)

In [32]:
clf = SVC(kernel = "rbf",C = 10000)
clf.fit(feature_train, label_train)
prediction = clf.predict(feature_test)    

In [40]:
num = 0
for i in range(len(prediction)):
    if prediction[i] == 1:
        num += 1

In [41]:
print num


877
